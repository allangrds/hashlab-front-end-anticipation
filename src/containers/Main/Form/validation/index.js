import maskedValueToCents from '../../../../utils/helpers/maskedValueToCents'

function validation (formValues) {
  const {
    amount,
    installments,
    mdr,
  } = formValues
  const errors = {}

  if (!amount) {
    errors.amount = 'Este campo é obrigatório'
  } else if (maskedValueToCents(amount) < 1000) {
    errors.amount = 'Valor muito baixo'
  } else if (maskedValueToCents(amount) > 100000000) {
    errors.amount = 'Valor muito alto'
  }

  if (!installments) {
    errors.installments = 'Este campo é obrigatório'
  } else if (installments < 1 || installments > 12) {
    errors.installments = 'Parcela inválida'
  }

  if (!mdr) {
    errors.mdr = 'Este campo é obrigatório'
  } else if (mdr < 0) {
    errors.mdr = 'MDR inválido'
  }

  return errors
}

export default validation
