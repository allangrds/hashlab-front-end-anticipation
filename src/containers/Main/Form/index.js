import React, { useState } from 'react'
import PropTypes from 'prop-types'

import {
  Button,
  Input,
} from '../../../components'
import styles from './styles.module.css'
import validation from './validation'
import maskedValueToCents from '../../../utils/helpers/maskedValueToCents'

const Form = ({
  loading,
  onClick,
}) => {
  const [formValues, setFormValues] = useState({
    amount: null,
    installments: null,
    mdr: null,
  })
  const [formErrors, setFormErrors] = useState({
    amount: null,
    installments: null,
    mdr: null,
  })

  function handleChange (input) {
    setFormValues({
      ...formValues,
      [input.name]: input.value,
    })
  }

  function handleSubmit (e) {
    e.preventDefault()

    const errors = validation(formValues)
    const noErrors = !errors.amount
      && !errors.installments
      && !errors.mdr

    if (noErrors) {
      onClick({
        ...formValues,
        amount: maskedValueToCents(formValues.amount),
      })
    }

    setFormErrors(errors)
  }

  return (
    <div className={styles.anticipation_simulatioon}>
      <h2 className={styles.title}>
        Simule sua Antecipação
      </h2>
      <form onSubmit={handleSubmit}>
        <div className={styles.input}>
          <Input
            onChange={handleChange}
            name="amount"
            type="money"
            label="Informe o valor da venda"
            required
            value={formValues.amount}
            errorMessage={formErrors.amount}
          />
        </div>
        <div className={styles.input}>
          <Input
            onChange={handleChange}
            name="installments"
            type="number"
            label="Em quantas parcelas"
            required
            tip="Máximo de 12 parcelas"
            min="1"
            max="12"
            value={formValues.installments}
            errorMessage={formErrors.installments}
          />
        </div>
        <div className={styles.input}>
          <Input
            onChange={handleChange}
            name="mdr"
            type="number"
            label="Informe o percentual de MDR"
            required
            min="1"
            value={formValues.mdr}
            errorMessage={formErrors.mdr}
          />
        </div>
        <Button
          disabled={loading}
          text="Simular"
        />
      </form>
    </div>
  )
}

Form.propTypes = {
  loading: PropTypes.bool.isRequired,
  onClick: PropTypes.func.isRequired,
}

export default Form
