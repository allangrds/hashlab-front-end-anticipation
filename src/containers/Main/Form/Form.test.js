import React from 'react'
import { render } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import Form from '.'

test('Form should render correctly', () => {
  const callback = jest.fn()

  const {
    container,
    getAllByText,
    getByLabelText,
    getByText,
    queryByText,
  } = render(
    <Form
      onClick={callback}
    />,
  )

  expect(callback).toHaveBeenCalledTimes(0)

  getByText('Simule sua Antecipação')

  const amountInput = getByLabelText('Informe o valor da venda *')
  const installmentsInput = getByLabelText('Em quantas parcelas *')
  const mdrInput = getByLabelText('Informe o percentual de MDR *')

  expect(amountInput.value.length).toEqual(0)
  expect(installmentsInput.value.length).toEqual(0)
  expect(mdrInput.value.length).toEqual(0)

  const button = getByText('Simular')
  userEvent.click(button)

  expect(getAllByText('Este campo é obrigatório').length).toEqual(3)

  userEvent.type(amountInput, '22,22')
  userEvent.type(installmentsInput, '1')
  userEvent.type(mdrInput, '0')

  userEvent.click(button)

  expect(queryByText('Este campo é obrigatório')).not.toBeInTheDocument()

  expect(callback).toHaveBeenCalledTimes(1)

  expect(container).toMatchSnapshot()
})
