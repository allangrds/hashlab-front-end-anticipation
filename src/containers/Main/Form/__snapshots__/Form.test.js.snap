// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should render correctly 1`] = `
<div>
  <div
    class="anticipation_simulatioon"
  >
    <h2
      class="title"
    >
      Simule sua Antecipação
    </h2>
    <div
      class="input"
    >
      <fieldset
        class="fieldset"
      >
        <label
          class="label"
          for="amount"
        >
          Informe o valor da venda
           
          *
        </label>
        <input
          class="input"
          id="amount"
          type="text"
          value="R$22,22"
        />
        <p
          class="tip"
        />
      </fieldset>
    </div>
    <div
      class="input"
    >
      <fieldset
        class="fieldset"
      >
        <label
          class="label"
          for="installments"
        >
          Em quantas parcelas
           
          *
        </label>
        <input
          class="input"
          id="installments"
          max="12"
          min="1"
          name="installments"
          type="number"
          value="1"
        />
        <p
          class="tip"
        >
          Máximo de 12 parcelas
        </p>
      </fieldset>
    </div>
    <div
      class="input"
    >
      <fieldset
        class="fieldset"
      >
        <label
          class="label"
          for="mdr"
        >
          Informe o percentual de MDR
           
          *
        </label>
        <input
          class="input"
          id="mdr"
          min="1"
          name="mdr"
          type="number"
          value="0"
        />
        <p
          class="tip"
        />
      </fieldset>
    </div>
    <button
      class="button"
      type="button"
    >
      Simular
    </button>
  </div>
</div>
`;
